colbert/index_updater.py

```diff
@@ -0,0 +1,178 @@
+import os
+import ujson
+import torch
+import numpy as np
+import tqdm
+
+from colbert.search.index_loader import IndexLoader
+from colbert.indexing.index_saver import IndexSaver
+from colbert.indexing.collection_encoder import CollectionEncoder
+
+from colbert.utils.utils import lengths2offsets, print_message, dotdict, flatten
+from colbert.indexing.codecs.residual import ResidualCodec
+from colbert.indexing.utils import optimize_ivf
+from colbert.search.strided_tensor import StridedTensor
+from colbert.modeling.checkpoint import Checkpoint
+from colbert.utils.utils import print_message, batch
+from colbert.data import Collection
+from colbert.indexing.codecs.residual_embeddings import ResidualEmbeddings
+from colbert.indexing.utils import optimize_ivf
+
+class IndexUpdater:
+    
+    def __init__(self, config, searcher, checkpoint=None):
+        self.config = config
+        self.searcher = searcher
+        if checkpoint:
+            self.checkpoint = Checkpoint(checkpoint, config)
+            self.encoder = CollectionEncoder(config, self.checkpoint)
+        self.index_path = searcher.index
+        self._load_disk_ivf()
+        
+        
+    def remove(self, pids, persist_to_disk=False):
+        for pid in pids:
+            print(f"removing {pid}...")
+            self._remove_pid_from_ivf(pid, persist_to_disk)
+            if persist_to_disk:
+                self._load_metadata()
+                self._remove_passage_from_disk(pid)
+    
+    def add(self, pids, persist_to_disk=False):
+        if not self.checkpoint:
+            raise ValueError("No checkpoint was provided at IndexUpdater initialization.")
+        # TODO: implement the add-passage functionalities
+        
+    def _load_disk_ivf(self):
+        print_message(f"#> Loading IVF...")
+
+        if os.path.exists(os.path.join(self.index_path, "ivf.pid.pt")):
+            ivf, ivf_lengths = torch.load(os.path.join(self.index_path, "ivf.pid.pt"), map_location='cpu')
+        else:
+            assert os.path.exists(os.path.join(self.index_path, "ivf.pt"))
+            ivf, ivf_lengths = torch.load(os.path.join(self.index_path, "ivf.pt"), map_location='cpu')
+            ivf, ivf_lengths = optimize_ivf(ivf, ivf_lengths, self.index_path)
+
+        self.curr_ivf = ivf
+        self.curr_ivf_lengths = ivf_lengths
+        
+    def _remove_pid_from_ivf(self, pid, persist_to_disk):
+        new_ivf = []
+        new_ivf_lengths = []
+        runner = 0
+        for l in self.curr_ivf_lengths.tolist():
```

Comment:
Nit: Could you replace `l` with `length`? The `l` looks too similar to a 1 so it's a bit confusing